Sources/BaggageContext/Context.swift

```diff
@@ -0,0 +1,287 @@
+///
+/// - SeeAlso: `Baggage` from the Baggage module.
+/// - SeeAlso: `Logger` from the SwiftLog package.
+public struct Context: ContextProtocol {
+    /// The `Baggage` carried with this context.
+    /// It's values will automatically be made available to the `logger` as metadata when logging.
+    ///
+    /// Baggage values are different from plain logging metadata in that they are intended to be
+    /// carried across process and node boundaries (serialized and deserialized) and are made
+    /// available to instruments using `swift-distributed-tracing`.
+    public var baggage: Baggage
+
+    public var logger: Logger {
+        get {
+            return self._logger.with(self.baggage)
+        }
+        set {
+            self._logger = newValue
+        }
+    }
+
+    private var _logger: Logger
+
+    public init(baggage: Baggage, logger underlying: Logger) {
+        self.baggage = baggage
+        self._logger = underlying
+    }
+
+    public var asBaggageContext: Context {
+        return self
+    }
+}
+
+// ==== ----------------------------------------------------------------------------------------------------------------
+// MARK: Context Protocol
+
+/// The `ContextProtocol` MAY be adopted by specific "framework contexts" such as e.g. `CoolFramework.Context` in
+/// order to allow users to pass such context directly to libraries accepting any context.
+///
+/// This allows frameworks and library authors to offer APIs which compose more easily.
+/// Please refer to the "Reference Implementation" notes on each of the requirements to know how to implement this protocol correctly.
+///
+/// ### Relation to `BaggageProtocol`
+/// The `ContextProtocol` does NOT extend `BaggageProtocol`.
+///
+/// A custom context implementation however `MAY` extend both `ContextProtocol` and `BaggageProtocol`.
+/// Doing so enables a custom context to obtain the baggage protocol's capability of getting *and setting*
+/// properties on a baggage directly, like so: `myContext.traceID = "111-222"` without having to type out the baggage
+/// explicitly, as would normally be the case (`myContext.baggage.traceID = "111-222"`).
+///
+/// This difference in capabilities is purposefully designed as such, in order to allow context objects implement `ContextProtocol`
+/// without having to force them to allow _modifying_ the baggage. Some frameworks or libraries may choose to create immutable
+/// context objects, including the baggage they offer there. Users, when wanting to add more contextual information need then to
+/// perform the `asBaggageContext`, mutate the resulting context OR use the `with...` fluent-API functions to obtain a mutated context.
+///
+/// Please note that if a context does conform to `BaggageProtocol` as well, it will most-likely also offer all properties that
+/// users have provided via extensions on it - so be watchful of potential naming conflicts on variable names.
+public protocol ContextProtocol {
```

Comment:
I think this protocol should be called just `Context` (with the default becoming `DefaultContext`). It will appear in code far more often than any concrete implementation.

Reply:
I tend to agree with this approach

Reply:
Yeah that's an option -- the approach is exactly as from the previous feedback about `DefaultContext` etc.

We could go with:

- `protocol Context` the thing that should be put in APIs when accepting types
- `struct DefaultContext` the default impl that holds logger and baggage

I was a bit worried of taking the name with a protocol but I guess realistically it's all the various library contexts that will conform to this and the default one is more rare in "bigger apps".

Then usage is:

- All APIs everywhere should: `func hello(who: String, context: Context)`
- Frameworks/Libraries should: `struct Framework.Context: Context`
- Small apps that done use frameworks have to `.hello(who: "", context: DefaultContext.background.with(...))` etc.

Reply:
I would really expect other libraries to have something more concrete than just `Context` like how NIO has `HandlerContext`. Vapor should have a `RequestContext` and `ApplicationContext` potentially.

At work, we're investigating using this library as the foundation for a `RuntimeContext` or `EnvironmentContext` that we'll be passing through our app

Reply:
> I would really expect other libraries to have something more concrete than just Context like how NIO has HandlerContext. Vapor should have a RequestContext and ApplicationContext potentially.

Yeah, though one has to be careful with those APIs... 

Soon enough you'll start wanting this `Context` library to also carry event loops and random other things... This simply does not compose well if statically defined in the protocol, so it won't be. If you build a library you must accept the wides possible range of contexts: the `Context` (protocol), and yes this means that you may have to pass other values separately. We've hit this with the carriers which did not compose well in the end.

I really want to keep stressing the fact that this context is for context propagation as understood by tracing systems and similar, and it will not evolve to have many other requirements in the `protocol Context`.

Reply:
> At work, we're investigating using this library as the foundation for a RuntimeContext or EnvironmentContext that we'll be passing through our app

Cool, yes I think that's good and intended 👍 Your contexts can carry whatever extra information you want since it's your _app_ and you control it. And on library edges you'd pass the same for the minimum tracing case... Should work well enough 👍

Reply:
Right, I should've clarified what I was referring to: The naming of `Context` vs `ContextProtocol`